### PR TITLE
Make `getindex`/`length` the default interface

### LINKED ISSUE #57: circularity in AbstractDataContainer definitions
We currently have the following definitions for `AbstractDataContainer` where `getindex` falls back to `getobs`
```julia
abstract type AbstractDataContainer end

Base.getindex(x::AbstractDataContainer, i) = getobs(x, i)
Base.length(x::AbstractDataContainer) = numobs(x)
Base.size(x::AbstractDataContainer) = (length(x),)

Base.iterate(x::AbstractDataContainer, state = 1) =
    (state > length(x)) ? nothing : (x[state], state + 1)
Base.lastindex(x::AbstractDataContainer) = length(x)
```
and on the other end we have the generic fallback for `getobs`
```julia
getobs(x, i) = getindex(x, i)
numobs(x) = length(x)
```
I find this circularity a bit confusing and think it should be avoided. I suggest we change `AbstractDataContainer` to

```julia
abstract type AbstractDataContainer end

Base.iterate(x::AbstractDataContainer, state = 1) =
    (state > numobs(x)) ? nothing : (getobs(x, state), state + 1)
Base.lastindex(x::AbstractDataContainer) = numobs(x)
```

Then types inheriting from `AbstractDataContainer`:
- Can implement `getindex` if they want both the "indexing" interface and the "observable" interface.
- Implement just `getobs` if for some reason they don't want to expose an indexing interface
- Implement both `getobs` and `getindex` if the two interfaces serve different purposes (e.g. as with arrays)

As an addendum, let me remark that with the `getobs(x, i) = getindex(x, i)` fallback we are basically saying that we consider a *Dataset* any type implementing `getindex`, which is something that maybe we should document more. Should defining `getindex` be the recommended way for defining custom dataset types (even if not subtyping `AbstractDataContainer`)? 

@darsnack 
related to https://github.com/JuliaML/MLDatasets.jl/pull/96

### PR DESCRIPTION
**Pull request by darsnack**
This removes the `getindex`/`length` defaults from `AbstractDataContainer`. Instead, types implementing the interface should implement `Base.getindex` or `Base.length` first, and only implement `getobs`/`numobs` when the fallback definitions don't apply.

Fixes #57.

### CODE REVIEW

**Comment by lorenzoh** [comment]
I support this change 👍  Makes things simpler for newcomers, makes it easier to take advantage of types from other packages that already implement indexing, and one no longer needs to import (and export) `getobs` and `numobs` everywhere. Also the same interface as `torch.utils.data.Dataset` and familiarity is good.

**Review by CarloLucibello — APPROVED** [review]
